src/Resolver/Resolver.php

```diff
@@ -107,18 +110,33 @@ public function resolve($domain)
      * ```
      *
      * @param string $domain
-     * @return PromiseInterface Returns a promise which resolves with all record values on success or
+     * @return Promise\PromiseInterface Returns a promise which resolves with all record values on success or
      *     rejects with an Exception on error.
      */
     public function resolveAll($domain, $type)
     {
+        $this->nameserverStart++;
+
+        if ($this->nameserverStart >= $this->nameserverCount) {
+            $this->nameserverStart = 0;
+        }
+
         $query = new Query($domain, $type, Message::CLASS_IN);
         $that = $this;
 
-        return $this->executor->query(
-            $this->nameserver,
-            $query
-        )->then(function (Message $response) use ($query, $that) {
+        $promise = Promise\reject();
+
+        for ($i = 0; $i < $this->nameserverCount; $i++) {
+            $nameserver = $this->nameservers[$this->nameserverStart++];
+            $promise = $promise->then(null, function () use ($nameserver, $query, $that) {
+                return $that->executor->query(
+                    $nameserver,
+                    $query
+                );
+            });
```

Comment:
I don't really like queueing fallbacks eagerly. In most (rather non-exceptional) situations they are unnecessary and a waste of CPU cycles.

Instead a recursively used function which queues the same query for the next nameserver should be the preferred way.
However it's fine with me if the other maintainers think it's fine, I'm not blindly stuck on this.

Reply:
Oh it's also not correctly implemented. It's very possible to try to get a non-existent nameserver. Imagine `$this->nameserverStart` = 4 and `$this->nameServerCount` = 5. The counter will be increased to 4 + 5 and there you have your out of bound error.

Reply:
I think you're having an interesting discussion. I agree that sending queries to all servers at once should probably not be the default - but I can see some value in this strategy depending on the use case. As such, I wonder if it makes sense to move this logic somewhere else? For example, by preparing the `ExecutorInterface` and implement different strategies there?